### ISSUE TITLE
RFC: API Refactor

### ISSUE THREAD

**Issue by epage** [cobalt-org/liquid-rust#95]
# Goals

Goals
- Layered so people pay for what they get
- Better separation of internals vs API
- Do not get in the way of using liquid in parallel applications
- Avoid copies with the context
  - cobalt has to copy in the vars initially
  - cobalt has reused vars that would be nice to have overlayed each other rather than copied.
- Make it possible to support expressions for include, see #142

Secondary goals
- Caching
  - Cache can be reused for cobalt's layouts/templates
    - Should this be left to clients?
  - Caching of snippets to avoid reading/parsing them
    - Is caching of top-level items good enough?

# Proposal

- [x] ParserBuilder -> Parser (#149)
  - Takes in includes, tags, blocks, and filters
- [x] Parser -> Template (#149)
  - Takes in str
- [x] Template -> String (#149)
  - Takes in globals
- [x] Separate concerns (#149)
  - Top-level Parser, Template
  - Value
  - Interpreter
  - Syntax / Compile to interpreted state
- [x] Update interpreter / syntax to take in state by reference where applicable
- [x] Update interpreter / syntax to take in state by traits
  - Including allowing custom backing stores for `globals`
  - How well can `Context` be a trait when different tags (cycle, include, break, etc) need special support in it?
- [ ] Update `trait FilterValue` to advertise the required/optional params
  - Must still support custom handling
  - Allow centralizing of error processing
  - Future: allow named parameters
- [x] Value API
  - `Value` should be an enum of `Object`, `Array`, `Scalar`, and `Nil`
  - `Scalar` is a struct wrapping an enum to hide implementation details
  - `Scalar` can coerce between different data types
- [ ] Crate structure and APIs to connect them
  - [x] liquid-value
  - [x] liquid-interpreter
  - [x] liquid-compiler
  - [x] liquid
  - [ ] tags, blocks, and filters
    - Do these sit above liquid with some API (extension traits?) to make it easy to add them to a parser?
    - Or do these sit below liquid with it having built-in support for adding these?
    - Are they added by "liquid" or "not" (like they are now) or by category with features to enable extra featurs
    - Keep in mind I want to split out from cobalt liquid-codeblock and [liquid-markdown](https://github.com/cobalt-org/cobalt.rs/issues/342)
- [ ] Should default implementations of Traits be moved out to leaf layers?
- [x] Explore [alternatives to passing around contexts](https://github.com/rust-lang-nursery/futures-rfcs/pull/2#issuecomment-363923477)

# State of template APIs

## liquid-rust

- `LiquidOption` owns
  - blocks
  - tags
  - file system access
- `parse` takes
  - text
  - `LiquidOptions`
- `Template`
  - `render` takes
    - `Context`
- `Context`
  - filters (standard filters automatically added)
  - values
  - various internal execution context publicly available

Concerns
- Seems like it should be restructured as builders
- From the client perspective, it seems odd that bocks/tags are pre-`parse` while filters are on-`render`

## [tera](https://docs.rs/tera/0.10.2/tera/struct.Tera.html)

- `Tera` is a template collection
  - `new` parses a directory (can add files to it one at a time)
  - `register_filter` for adding them to the template collection
  - `register_tester` for adding new conditional evaluations
  - `render` takes a template identifier and a `Context`

**Comment by epage** [comment]
I think `Context` needs to be broken up into the immutable state the client passes in and the mutations that need to happen as part of executing liquid template code.

The immutable state should borrow rather than own the var that comes in.  In addition, a iterator of vars should be supported for treating as one without doing any copies.

See https://github.com/cobalt-org/cobalt.rs/issues/256 for an example of why that is needed.

**Comment by epage** [comment]
`ParserBuilder`
- Very similar to `LiquidOptions`, just without `pub` fields
- Would love to find a way to move `template_repository` out.
  - The problem is that currently, included files are parsed and inserted directly into the parsed output.  
  - Delaying this to `render` would require having pre-parsed files available somehow.
  - Or somehow allow on-demand parsing

`Parser`
- Owns the equivalent of LiquidOptions
- Has parse functions that output `Template`

`Template`
- New public type, separate from renderable
- No automatic filters being added

`Context`
- filters
- `Vec<&Value>`
- Implement a trait for looking up values in `Vec`

`State`
- `&Context`
- stack, interrupt, and cycles
- Implement a trait for looking up values in `Context` and `stack`

Possible optimization for `Context` and `Stack`
- Implement a trie for `*&Value` that are nested like `site.data.folder.file.*` to avoid iterating over 10 entries of data files in the same folder.

**Comment by epage** [comment]
@gmalette since you seem to be the main non-cobalt user of liquid, I'd value your thoughts on any API redesign, especially as it impacts `TemplateRepository` (see #90).

As you'll see noted above, I'm considering what the API should look like.

Assumptions (ie the case being optimized for)
- Snippets are generally used for reused parts across cobalt layouts (header, footer)
  - Included snippets `Template`s are small
  - Each layout will include 0-4 snippets
  - Generally the 0-4 snippets will be the same across an entire blog (hundreds of pages)

Requirements
- Snippets have access to all the same variables as the layout and page
- Snippets can arbitrarily include other snippets (including cycles as long as they eventually terminate)

Nice-to-have
- Enforce snippets are parsed with the same `Parser` as the relevant layout/page to ensure correctness

From that
- Snippets must be rendered at the same time as a layout or page and get the variables from them
- Don't need to worry about reducing the number of in-memory snippets because they are small
- Parse time of a snippet should be minimal
- disk IO could be the main factor to optimize for


- Ideally, cobalt would be able to parse an included file once and be done with it.
  - Besides saving on parsing time, this reduces disk IO which will probably be the bigger bottleneck
  - Current `TemplateRepository` could do this
- Ideally, the API enforces the same `Parser` being used between the document and snippit templates (
  - Current `TemplateRepository` can't do this
  - Granted, I don't think any of my alternatives could either...

Ideas

# `Parser` owns a `TemplateRepository`

- This is the current solution.
- Reduce disk io: cobalt could do some fancy implementation details for a `TemplateRepository` to cache the text in-memory
- Same parser: Yes
- Avoid `Box`: Can't without making everything generic both on `TemplateRepository` and its lifetime
- Can cache parsed snippets: no

# Pre-load `Parser` with snippet text

- Requires knowing all snippets a priori
  - This is fine for cobalt.  Some might be surprised that an in-work snippet failing would fail their site build
  - this might not work for @gmalette's case
- Reduce disk io: cobalt can walk the layout folder, read the text, and feed it into the parser.  The assumption being that there won't be extraneous snippets in the folder.
- Same parser: Yes
- Avoid `Box`: Yes
- Can cache parsed snippets: yes

# Pre-load `Context` with snippet text

Effectively the same as loading it in the parser, just delays the parsing

- Can cache parsed snippets: generally no unless a page reuses a snippet

# `TemplateRepository` is included in `Context`

Either the `Template` needs a `.clone()` of the `Parser` or they need to be pre-parsed which might have been done with a different parser

**Comment by epage** [comment]
I realized that things like preloading the parser still can't pre-parse because what would when a snippet recurses?  it will need a pre-parsed dependencies in there as well.

**Comment by epage** [comment]
For `Filter`s.  We box them when almost no filter needs that.

What if we had an `Either<(parts of a Filter), Box>` or two separate maps?  The `(parts of a filter)` could even include an `Object` to handle the 90% case of configurable filters.

Also, we might want to make the `Filter` trait more complex.  For one, all of the filters have to do their own error processing for required and optional parameters.  Instead it'd be nice if a filter advertised its params and the error processing was done for them.  We'd need to somehow support variable number of params.  This could also be reused for named argument support (see #92).

**Comment by epage** [comment]
`Context: Implement a trait for looking up values in Vec`

What if we made `Context` return an `Expression` that we could perform various operations on?  The downside would be that it would need to be boxed.  The upside is that `Context` could have alternative backends besides a `Vec<&Object>` and it could allow some more lazy operations, like indexing.

**Comment by gmalette** [comment]
In my case, I expect a lot more than 0-4 snippets (across all loaded files by the `TemplateRepository`, I can expect hundreds for a single layout. However I do lazy rendering, so that should be fine, so long as templates/snippets can be loaded in an abstract manner (`TemplateRepository`, right now).

Having `Context` have an abstract backend would be huge for me.

That being said, I've pretty much put the project on ice after semi-failing to do named args :(. I really appreciate you looking out for me, but at this point it's wiser to focus your efforts on your own needs.

**Comment by epage** [comment]
Conceptually, what state belongs on Parser vs Context?

If thinking about this from programming language build process
- TemplateRepository
  - when building your project, you configure the dependencies up front and that is reused across the entire project
  - But you can create extra files or snippets within your own project
  - Snippets are immutable across different templates being rendered
- filters
  - stdlib functions are effectively built into the language
  - custom functions can be provided though
  - liquid doesn't quite have custom functions the same way as a programming language, snippets are more equivelant
  - filters are immutable across different templates being rendered
- variables
  - variables are  unique for each template
- top level file
  - this is unique for each template.

So I'm tempted to move `filters` into the parser even though they aren't technically needed the same way as `TemplateRepository`.

**Comment by epage** [comment]
https://github.com/cobalt-org/liquid-rust/pull/195 got some improvements in

**Comment by epage** [comment]
- [ ] Make `Index` use `Cow`
- [ ] Make `include` be runtime rather than compile time.

**Comment by epage** [comment]
The ruby implementation keeps all tag-specific context in [`registers`](https://github.com/Shopify/liquid/blob/master/lib/liquid/tags/ifchanged.rb).  That might be a bit easier in a dynamic language but we should look into it.

Inspired by #211

**Comment by epage** [comment]
See https://github.com/cobalt-org/liquid-rust/issues/301
